src/clj/kixi/hecuba/data/measurements/calculations.clj

```diff
@@ -242,3 +244,119 @@
+                               :end (if (< next tarrifs-count)
+                                      (tc/to-date-time (:timestamp (nth tariffs next)))
+                                      (time/update-timestamp (t/now) {:hour 0 :minutes 0 :seconds 0
+                                                                      :milliseconds 0}))
+                               :tariff (-> item :profile_data :tariff)))) tariffs)))
+
+(defn match-tariff
+  "Finds tariff that matches period of a measurement."
+  [m tariffs]
+  (let [t (-> m :timestamp tc/to-date-time)]
+    (first (keep #(when (time/in-range? t (:start %) (:end %)) %) (tariff-ranges tariffs)))))
+
+(defn discounted-standing-charge
+  "Gets daily standing charge and total annual lump-sum discount
+  and calculates discounted standing charge.
+  Returns a numerical value."
+  [daily-standing-charge annual-lump-sum-discount]
+  (float (/ (- (* daily-standing-charge 365) annual-lump-sum-discount) 365)))
+
+(defmulti apply-tariff
+  "Processes one day's worth of data (differenceSeries),
+  works out unit tariff from the tariff map provided and
+  multiplies it by the consumption for each period.
+  Returns a sequence of expenditure measurements."
+  (fn [tariff _]
+    (cond
+      (:cost-per-on-peak-kwh tariff) :on-off-peak
+      :else :simple)))
+
+(defmethod apply-tariff :on-off-peak [tariff measurements]
+  (let [{:keys [cost-per-kwh daily-standing-charge
+                annual-lump-sum-discount]} tariff
+        standing-charge (discounted-standing-charge daily-standing-charge
+                                                    annual-lump-sum-discount)]))
+
+(defmethod apply-tariff :simple [tariff measurements]
+  (let [{:keys [cost-per-kwh daily-standing-charge
+                annual-lump-sum-discount]} tariff
+        standing-charge (discounted-standing-charge daily-standing-charge
+                                                    annual-lump-sum-discount)
+        total-consumption (apply + (map :value (filter #(number? (:value %)) measurements)))]
+    (-> total-consumption
+        (* cost-per-kwh)
+        (+ standing-charge))))
+
+(defn calculate-tariff-batch
+  "Gets store and sensor information, start
+  and end dates and a list of tariffs.
+  Retrieves a day's worth of data, matches appropriate tariff
+  and calculates expenditure.
+  Inserts result of that calculation into the database."
+  [store sensor start-date end-date tariffs]
+  (db/with-session [session (:hecuba-session store)]
+    (let [device_id        (:device_id sensor)
+          sensor_id        (:sensor_id sensor)
+          month            (time/get-month-partition-key start-date)
+          where            [[= :device_id device_id] [= :sensor_id sensor_id] [= :month month]
+                            [>= :timestamp start-date] [< :timestamp end-date]]
+          measurements     (measurements/fetch-measurements store where)
```

Comment:
I'm not well up on this code, but it seems likely that there are some well known predicates that we'd re-use, which I think is what you're suggesting. something like (untested):
```clojure
...
where (concat (matches-sensor sensor)
              (in-range start-date end-data)]
...
```
with 
```clojure
(defn matches-sensor({:keys [device_id sensor_id})
     [[= :device_id device_id] [= :sensor_id sensor_id]])
```